libcanard/canard_dsdl.c

```diff
@@ -59,22 +59,31 @@ CANARD_PRIVATE size_t chooseMin(size_t a, size_t b)
     return (a < b) ? a : b;
 }
 
-/// The algorithm was originally designed by Ben Dyer for Libuavcan v0:
-/// https://github.com/UAVCAN/libuavcan/blob/ba696029f9625d7ea3eb00/libuavcan/src/marshal/uc_bit_array_copy.cpp#L12-L58
-/// This version is modified for v1 where the bit order is the opposite.
-/// If both offsets and the length are byte-aligned, the algorithm degenerates to memcpy().
-/// The source and the destination shall not overlap.
-CANARD_PRIVATE void copyBitArray(const size_t         length_bit,
-                                 const size_t         src_offset_bit,
-                                 const size_t         dst_offset_bit,
-                                 const uint8_t* const src,
-                                 uint8_t* const       dst);
-CANARD_PRIVATE void copyBitArray(const size_t         length_bit,
-                                 const size_t         src_offset_bit,
-                                 const size_t         dst_offset_bit,
-                                 const uint8_t* const src,
-                                 uint8_t* const       dst)
+CANARD_PRIVATE size_t getBitCopySize(const size_t  buf_size_bytes,
+                                     const size_t  offset_bit,
+                                     const size_t  requested_length_bit,
+                                     const uint8_t value_length_bit);
+CANARD_PRIVATE size_t getBitCopySize(const size_t  buf_size_bytes,
+                                     const size_t  offset_bit,
+                                     const size_t  requested_length_bit,
+                                     const uint8_t value_length_bit)
+{
+    const size_t buf_size_bit  = buf_size_bytes * BYTE_WIDTH;
+    const size_t remaining_bit = buf_size_bit - chooseMin(buf_size_bit, offset_bit);
+    return chooseMin(remaining_bit, chooseMin(requested_length_bit, value_length_bit));
+}
+
+// --------------------------------------------- PUBLIC API - BIT ARRAY ---------------------------------------------
+
+void canardDSDLCopyBits(const size_t         length_bit,
```

Comment:
It'd be nice to document this public API. Most of it is obvious but it's not clear what `length_bit` is the length of; the src array or the dst array?

This also uses size_t which is intended for byte-oriented lengths and isn't appropriate for bit-oriented offsets. This is more of a systemic comment though and doesn't require an action for this PR.

Reply:
Also, document what errors can occur and their effect. What happens if src is null? dst? What happens if there is a buffer overrun? What happens if src and dst overlap?